### ISSUE TITLE
Contracting 'mixedIQTensor' and 'IQTensor'

### ISSUE THREAD

**Issue by maximedupont** [ITensor/ITensor#103]
Hi,

To be consistent in the code I think that when defining mixedIQTensor operators (`Sx`, `ISy` and `Sy`) in the files _itensor/mps/sites/spinhalf.h_ and _itensor/mps/sites/spinone.h_ it should be `Op = mixedIQTensor(dag(s),sP);` rather than  `Op = mixedIQTensor(s,sP);`.

Now if one wants to contract one of these operators with an IQTensor he gets the following error:

> terminate called after throwing an instance of 'ITError'
>   what():  doTask not defined for task Contract and storage types QDenseReal QMixed<Real>
> Aborted

Unless he converts the IQTensor first to an ITensor using `toITensor()`. And now that we are working with ITensors the first remark is not relevant...

The question is that I'm not sure the code is designed work **only if** the IQTensor is explicitely converted to an ITensor before contracting. Maybe this can be directly done/implemented in the ITensor library?

Best,
Maxime

**Comment by emstoudenmire** [comment]
Hi Maxime,
Thanks for reporting this. However, the behavior you encountered is actually the intended design, confusing as it may be. 

Here is the logic behind mixedIQTensor:
1. The SiteSet::op method can only have one return type, so we made it IQTensor because IQTensor is convertible to ITensor.
2. We want to be able to make operators that don't conserve Sz. For example, to make Hamiltonians with magnetic fields in the x- or y-direction, for example.
3. So we came up with mixedIQTensor which is an object of type IQTensor, but which is "morally" just an ITensor in that it does not have a well defined quantum-number "flux" i.e. it does not change Sz in a definite way. The purpose of a mixedIQTensor is just to be converted to ITensor immediately.

So in a nutshell, one should not use these operators in a quantum-number conserving code (one that is built around IQTensor) because they don't conserve quantum numbers. There are alternatives to using Sx or Sy that do conserve quantum numbers. For example, when making the Heisenberg Hamiltonian, write it in terms of "S+" and "S-" instead, as these transform the total Sz in a well-defined way, unlike "Sx" or "Sy".

What was the use-case you were trying to use these operators for?

**Comment by maximedupont** [comment]
Thanks for the quick answer.

I wanted to compute correlation functions `<Sx_i Sx_j>` within a specific Sz_tot sector. In this context, having to convert explicitly all the IQTensors to ITensors when contracting them seemed "unnatural".

For example, using the following code (adapted to version 2) [http://itensor.org/docs.cgi?page=formulas/correlator_mps](http://itensor.org/docs.cgi?page=formulas/correlator_mps) with operators that don't conserve Sz implies replacing each `psi.A(n)` to `toITensor(psi.A(n))`.

**Comment by emstoudenmire** [comment]
Hi Maxime,
This is a part of our design that we might improve in the future by making two kinds of SiteSet objects which either return only ITensors or IQTensors. But this would likely be a breaking change, so would have to wait until a new major release of the library.

Regarding your specific case, you should rewrite the operators `Sx_j` as `Sx_j = 0.5 * (S+_j + S-_j)` and rewrite your correlation function in terms of those operators. This way each intermediate tensor in your calculation has a well defined quantum number flux (i.e. transforms in a well-defined way under z-axis rotations) which is the property you want when dealing with IQTensors (it allows the block-sparse structure of the tensors to be preserved).

Even if we changed the design as I mentioned above, in the new design you actually wouldn't be able to request the "Sx" operator at all when working with IQTensors, so you would still get an error for the code you are trying to write - the error would just occur earlier and have a more helpful error message.

Miles

**Comment by maximedupont** [comment]
Hi Miles,
Thank you for your answers. :-)

I'm closing the ticket!
